src/mlpack/methods/ann/layer/lstm_impl.hpp

```diff
@@ -18,430 +18,296 @@
+  if (!this->HasPreviousStep())
   {
-    cell.cols(forwardStep, forwardStep + batchStep) =
-        inputGateActivation.cols(forwardStep, forwardStep + batchStep) %
-        hiddenLayerActivation.cols(forwardStep, forwardStep + batchStep);
+    cell.slice(this->CurrentStep()) =
+        inputGateActivation.slice(this->CurrentStep()) %
+        hiddenLayerActivation.slice(this->CurrentStep());
   }
   else
   {
-    cell.cols(forwardStep, forwardStep + batchStep) =
-        forgetGateActivation.cols(forwardStep, forwardStep + batchStep) %
-        cell.cols(forwardStep - batchSize, forwardStep - batchSize + batchStep)
-        + inputGateActivation.cols(forwardStep, forwardStep + batchStep) %
-        hiddenLayerActivation.cols(forwardStep, forwardStep + batchStep);
+    cell.slice(this->CurrentStep()) =
+        forgetGateActivation.slice(this->CurrentStep()) %
+        cell.slice(this->PreviousStep()) +
+        inputGateActivation.slice(this->CurrentStep()) %
+        hiddenLayerActivation.slice(this->CurrentStep());
   }
 
-  outputGate.cols(forwardStep, forwardStep + batchStep) = input2GateOutputWeight
-      * input + output2GateOutputWeight * outParameter.cols(
-      forwardStep, forwardStep + batchStep) + cell.cols(forwardStep,
-      forwardStep + batchStep).each_col() % cell2GateOutputWeight;
-
-  outputGate.cols(forwardStep, forwardStep + batchStep).each_col() +=
-      input2GateOutputBias;
-
-  outputGateActivation.cols(forwardStep, forwardStep + batchStep) = 1.0 /
-      (1 + arma::exp(-outputGate.cols(forwardStep, forwardStep + batchStep)));
-
-  cellActivation.cols(forwardStep, forwardStep + batchStep) =
-      arma::tanh(cell.cols(forwardStep, forwardStep + batchStep));
+  outputGate = input2GateOutputWeight * input +
+      cell.slice(this->CurrentStep()).each_col() % cell2GateOutputWeight;
+  if (this->HasPreviousStep())
+  {
+    outputGate +=
+        output2GateOutputWeight * outParameter.slice(this->PreviousStep());
+  }
+  outputGate.each_col() += input2GateOutputBias;
 
-  outParameter.cols(forwardStep + batchSize,
-      forwardStep + batchSize + batchStep) =
-      cellActivation.cols(forwardStep, forwardStep + batchStep) %
-      outputGateActivation.cols(forwardStep, forwardStep + batchStep);
+  outputGateActivation.slice(this->CurrentStep()) =
+      1.0 / (1.0 + arma::exp(-outputGate));
 
-  output = OutputType(outParameter.memptr() +
-      (forwardStep + batchSize) * outSize, outSize, batchSize, false, false);
+  cellActivation.slice(this->CurrentStep()) =
+      arma::tanh(cell.slice(this->CurrentStep()));
 
-  cellState = OutputType(cell.memptr() +
-      forwardStep * outSize, outSize, batchSize, false, false);
+  // There's a bit of an issue here: we need to preserve the output for the next
+  // time step, but we also need to set `output` to that.  Unfortunately for now
+  // we make a copy, but it's possible that we could instead use an alias here,
+  // or have `outParameter` hold a collection of aliases.
```

Comment:
Supporting a collection of aliases sounds like a good idea to me.

Reply:
Yeah---however I'd rather save it for a later TODO.  We can add this to the list of issues to open later. :)